man/lseen.man
.\" 3.2
.\"  /*******************************************************************\
.\"   * Copyright (C) 1995 Lars J Aas <larsa@colargol.stud.idb.hist.no> *
.\"   * based on documentation by C J Needham <col@imdb.com> 1995,      *
.\"   * permission is granted by the authors to freely distribute       *
.\"   *                        providing no fee of any kind is charged. *
.\"  \*******************************************************************/
.\"
.TH LSEEN 1 "RELEASEDATE" " " "The Internet Movie Database vVERSIONNUMBER"
.SH NAME
lseen \- for use in conjunction with movie ratings report voting support scripts.
.SH SYNOPSYS
lseen
.SH DESCRIPTION
This script is for use in conjunction with my
movie ratings report voting support scripts.
This program counts/sorts/formats lists, extracting films you have
seen based on data from your votelog file.
.PP
If you don't use voting support scripts, they are available via
anonymous ftp from cathouse.org in the directory /pub/cathouse/movies/database/tools/voting_scripts.Z.
.PP
If you'd like a copy of your votelog file from the movie ratings
database, contact col@imdb.com via email, and he'll send it
to you.
.SH OPTIONS
There are
.BR database-options ,
.B output-options
and
.BR control-options .
.SS DATABASE-OPTIONS
.TP
.B \-acr
Use actors database.
.TP
.B \-acs
Use actresses database.
.TP
.B \-dir
Use directors database.
.TP
.B \-write
Use writers database.
.TP
.B \-comp
Use composers database.
.TP
.B \-cine
Use cinematographers database.
.TP
.B \-edit
Use editors database.
.TP
.B \-prodes
Use production-designers database.
.TP
.B \-costdes
Use costume-designers database.
.TP
.B \-prdcr
Use producers database.
.TP
.B \-misc
Use miscellaneous database.
.SS OUTPUT-OPTIONS
.TP
.B \-us
unsorted; don't sort data by total movies you have seen.
.TP
.B \-yr
Add the year of release to each title, and sort chronologically.
.TP
.B \-yru
Add the year of release, but leave unsorted.
.TP
.B \-mrr
Add movie ratings report info.
.TP
.B \-smrr
Add movie ratings report, sorted by average ratings.
.TP
.B \-vmrr
Add movie ratings report, sorted by number of votes.
.SS CONTROL-OPTIONS
.TP
.B \-min <val>
Limit output to people you've seen at least
.BR <val> of their movies.
.SH EXAMPLES
.TP
Count directors:
.B
lseen \-dir
.TP
Count actors and display ratings report and year of release info for
each movie:
.B
lseen \-acr \-mrr \-yr


docs/ChangeLog
-------Version 3.4 Released---------------------

Sun Mar 29 22:40:37 BST 1998

 * biographies list offset fields increased to 4 bytes following overflow
 * various constants updated

-------Version 3.3c Released---------------------

Sat Apr 26 06:51:19 BST 1997

 * added support for crew completion list
 * added support for distributors list
 * adapted mkdbs for more robust processing of filmography lists
 * added support for new business list tag

-------Version 3.3b Released---------------------

Sat Mar 15 08:09:50 GMT 1997

 * added support for ISO-Latin-1

-------Version 3.3 Released---------------------

Wed Oct 16 22:28:54 BST 1996

 * added support for AKA attributes
 * added support for writers ordering
 * moved misc section to after case in title output
 * added support for business list
 * added support for laserdisc list
 * fixed tagline bug
 * diplays URL of web page for name/title

-------Version 3.2h Released---------------------

Tue Oct  8 14:24:21 BST 1996

 * added support for special effects companies list
 * updated various system constants
 * added new movie link types 'edited into' & 'edited from'
 * added extra checks to avoid title/attribute key file corruption in mkdbs.c

-------Version 3.2f Released---------------------

Mon Jun 24 22:04:46 GMT 1996

 * added support for language list

-------Version 3.2e Released---------------------

Sat Mar 16 08:09:39 GMT 1996

 * added support for alternate versions list
 * added bio keys IT & AT

-------Version 3.2d Released---------------------

Tue Aug 15 23:23:47 GMT 1995

 * switched TitleID to unsigned long to handle >65K titles
 * fixed bug in display of literature and genre data
 * added bio key WN

-------Version 3.2c Released---------------------

Sun May 14 23:00:14 GMT 1995

 * raised MAXFILMOGRAPHIES constant to handle larger no. of entries

-------Version 3.2 Released---------------------

Sun Jan 29 22:40:08 GMT 1995

 * added full support for title searching without year/TV/V info

Tue Jan  3 23:00:45 GMT 1995

 * replaced 'biodefs' file
 * fixed bug when looking for common titles with people using -name

Mon Jan  2 21:52:08 GMT 1995

 * log file support
 * better separation of search/display code

Wed Dec 14 21:43:57 GMT 1994

 * fixed bug in 'title' for miscellaneous section
 * added support for the movie links list

Tue Dec 13 22:06:44 GMT 1994
 
 * noted when cast in credits output
 * fixed name ordering problem for names with (I)/(II) marker
 * display Actor/Actress in list -name where appropriate

Wed Dec  7 21:19:19 GMT 1994

 * added support for tag lines & complete cast list
 * put spaces around "." in cast results
 * standardized number of dashes printed out in 'title' / 'list'

Mon Dec  5 21:19:42 GMT 1994

 * multi-country productions displayed on same line

Wed Nov 30 18:14:16 GMT 1994

 * switched to new IMDB license
 * updated 'lfetch' to get new lists
 * fixed "No plot summaries message"
 * simplified MRR processing in mkdb
 * added support for locations & release date lists
 * switched 'Reviews' marker for clarity
 * added support for technical & literature lists

Mon Oct 31 21:42:03 GMT 1994
 
-------Version 3.1 Released---------------------

Mon Oct 17 19:04:26 GMT 1994
 
 * 3.1 code frozen
 * minor bug fixes with lindex

Sun Oct  9 18:37:14 GMT 1994

 * added support for countries list

Tue Oct  4 21:27:53 GMT 1994

 * added support for cast ordering info

Sun Oct  2 23:02:59 GMT 1994

 * added support for color info and sound mix lists
 * added constraint options to 'list'
 * standardized list options

Wed Sep 28 22:22:56 GMT 1994

 * re-ordered title info printing plus separated genres list
 * print genres on one line

Mon Sep 26 21:53:47 GMT 1994

 * added support for production companies list

Sun Sep 25 22:02:07 GMT 1994

 * added constraint options to 'lindex'

Tue Sep  6 20:09:52 GMT 1994

 * added support for certificates & genres list
 * fixed mkballot bug with -ar option

Tue Aug 23 20:08:04 GMT 1994

 * added support for miscellaneous filmographies list

Tue May 24 19:13:47 GMT 1994

 * moviedb-3.0 completed

Thu May 19 22:35:12 GMT 1994

 * switched to merged character names scheme

Tue May 17 21:09:45 GMT 1994

 * new 'mkdb' eliminating need for large array to hold name keys
 * removed duplicate aka titles in 'title -s -aka' queries

Wed May 11 22:11:59 GMT 1994

 * made upperCase a macro for increased speed
 * switched to Phil's new Makefiles

Wed May  4 17:46:08 GMT 1994

 * mapped case of titles/names to match key files
 * added support for running times list
 * lguide error messages

Tue May  3 21:38:30 GMT 1994
 
 * added support for Phil's imoviedb package
 * expanded compression support

Thu Apr 28 19:51:52 GMT 1994

 * finished rewrite to use new database representation

-------Version 2.9 Released---------------------

Wed Mar 16 21:39:40 GMT 1994

 * added support for producers list
 * added support for soundtracks list
 * added support for generic aka lists

Tue Mar 15 23:47:46 GMT 1994

 * replaced -crazy and -triv options with -trivia (also includes goofs & quotes)
 * added support for quotes list
 * added support for goofs list
 * handled -full error messages correctly
 * correct conditions for printing of missing plot summaries message
 * fixed lindex to search all lists when no list options given
 * -mrr no longer implies -m

Mon Mar 14 21:48:21 GMT 1994

 * new version of check-for-program
 * passed $MAKE variable to make subprocesses
 * lfetch modified to match full filenames
 * added 'keepchars' option for CLEANLISTS

-------Version 2.8 Released---------------------

Wed Dec  1 21:58:21 GMT 1993

 * man pages added to package
 * LaTeX documentation added to package

Sat Nov 27 20:53:40 GMT 1993

 * 2.8 code finished!

Thu Nov 25 18:17:00 GMT 1993

 * ported lseen to C

Tue Nov 23 19:16:48 GMT 1993

 * added gendb -votes option

Mon Nov 22 23:54:58 GMT 1993

 * ported lformat to C
 * removed yrf.c file

Wed Nov 17 22:36:44 GMT 1993


src/lseen.c
/*============================================================================
 *
 *  Program: lseen.c 
 *
 *  Version: 3.18
 *
 *  Purpose: track movies seen
 *
 *  Author:  C J Needham <col@imdb.com>     
 *
 *  Copyright (c) 1996-2002 The Internet Movie Database Inc.
 *
 *  Permission is granted by the copyright holder to distribute this program
 *  is source form only, providing this notice remains intact, and no fee
 *  of any kind is charged. This software is provided as is and there are no 
 *  warranties, expressed or implied.
 * 
 *  Options:
 *
 *        -acr    track actors database
 *        -acs    track actress database
 *        -dir    track directors database
 *      -write    track writers database
 *       -comp    track composers database
 *       -cine    track cinematographers database
 *       -edit    track editors database
 *     -prodes    track production designers database
 *    -costdes    track costume designers database
 *      -prdcr    track producers database
 *       -misc    track miscellaneous database
 *
 *    -mrr        add movie ratings report info to results
 *    -smrr       add movie ratings report info & sort by ratings
 *    -vmrr       add movie ratings report info & sort by votes
 *
 *    -yru        add year of release info
 *    -yr         add year of release info & sort chronologically
 *
 *    -min <val>  only report on people where you've seen <val> or more of
 *                 their movies
 *
 *============================================================================
 */

#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include "moviedb.h"
#include "dbutils.h"
#include "years.h"
#include "ratings.h"

#define LSEEN_USAGE1 "usage: lseen [-acr|-acs|-dir|-write|-comp|-cine|-edit|-prodes|-costdes|-prdcr"
#define LSEEN_USAGE2 "              -misc] [-mrr|-smrr|-vmrr] [-yr|-yru] [-us] [-min <value>]"


void displayLseenYears ( struct formatRec listData [], int count ) 
{
  int i ;

  (void) printf ( "%s (%d)\n\n", listData [ 0 ] . name, count ) ;

  for ( i = 0 ; i < count ; i++ )
  {
    (void) printf ( "   " ) ;
    if ( listData [ i ] . year > 0 )
      displayTitleYear ( listData [ i ] . title, listData [ i ] . year ) ;
    else
      (void) printf ( "%s", listData [ i ] . title ) ;
    if ( listData [ i ] . attr != NULL )
      (void) printf ( " %s\n", listData [ i ] . attr ) ;
    else
      (void) printf ( "\n" ) ;
  }
  (void) printf ( "\n" ) ;
}


void displayLseenRatings ( struct formatRec listData [], int count ) 
{
  int i ;

  (void) printf ( "%s (%d)\n\n", listData [ 0 ] . name, count ) ;

  for ( i = 0 ; i < count ; i++ )
  {
    if ( listData [ i ] . mrr . votes > 0 )
      (void) printf ( "      %s%8d   %2.1f  ", listData [ i ] . mrr . distribution,
                 listData [ i ] . mrr . votes, listData [ i ] . mrr . rating ) ;
    else
      (void) printf ( "                                " ) ;

    if ( listData [ i ] . year > 0 )
      displayTitleYear ( listData [ i ] . title, listData [ i ] . year ) ;
    else
      (void) printf ( "%s", listData [ i ] . title ) ;
    if ( listData [ i ] . attr != NULL )
      (void) printf ( " %s\n", listData [ i ] . attr ) ;
    else
      (void) printf ( "\n" ) ;
  }
  (void) printf ( "\n" ) ;
}


void displayLseenData ( struct formatRec listData [], int count, int mrropt, int yropt ) 
{
  if ( mrropt != NONE )
    displayLseenRatings ( listData, count ) ;
  else if ( yropt != NONE )
    displayLseenYears ( listData, count ) ;
  else
    (void) printf ( "%s (%d)\n", listData [ 0 ] . name, count ) ;
}


int mrrTitleKeyLseenCompare ( struct mrrDbRec *search, struct mrrDbRec *m1 )
{
  return ( search -> titleKey - m1 -> titleKey ) ;
}


int yearTitleKeyLseenCompare ( struct titleDbRec *search, struct titleDbRec *t1 )
{
  return ( search -> titleKey - t1 -> titleKey ) ;
}


void mapLseenData ( struct formatRec listData [], int count, FILE *titleKeyFp, FILE *titleIndexFp, FILE *attrKeyFp, FILE *attrIndexFp )
{
  int i ;

  for ( i = 0 ; i < count ; i++ )
  {
    listData [ i ] . title = mapTitleKeyToText ( listData [ i ] . titleKey, titleKeyFp, titleIndexFp ) ;
    listData [ i ] . attr = mapAttrKeyToText ( listData [ i ] . attrKey, attrKeyFp, attrIndexFp ) ;
  }
}


void sortLseenData ( struct formatRec listData [], int count, int mrropt, int yropt, struct titleDbRec *years, TitleID nyears, struct mrrDbRec *ratings, TitleID nratings ) 
{
  struct mrrDbRec *mrec, mSearchRec ;
  struct titleDbRec *trec, tSearchRec ;
  int i ;

  if ( yropt != NONE )
  {
    for ( i = 0 ; i < count ; i++ )
    {
      tSearchRec . titleKey = listData [ i ] . titleKey ;
      if ( ( trec = bsearch ( (void*) &tSearchRec, (void*) years, (size_t) nyears, sizeof ( struct titleDbRec ), (int (*) (const void*, const void*)) yearTitleKeyLseenCompare ) ) != NULL )
        listData [ i ] . year = trec -> year ;
      else
        listData [ i ] . year = 0 ;
    }
  }

  if ( mrropt != NONE )
  {
    for ( i = 0 ; i < count ; i++ )
    {
      mSearchRec . titleKey = listData [ i ] . titleKey ;
      if ( ( mrec = bsearch ( (void*) &mSearchRec, (void*) ratings, (size_t) nratings, sizeof ( struct mrrDbRec ), (int (*) (const void*, const void*)) mrrTitleKeyLseenCompare ) ) != NULL )
        listData [ i ] . mrr = mrec -> mrr ;  
      else
      {
        listData [ i ] . mrr . votes = 0 ;  
        listData [ i ] . mrr . rating = 0.0 ;  
      }
    }
  }

  if ( mrropt == NONE && yropt == YR ) 
    qsort ( (void*) listData, (size_t) count, sizeof ( struct formatRec ), (int (*) (const void*, const void*)) yearLformatSort ) ;
  else if ( mrropt == MRR && yropt == YR ) 
    qsort ( (void*) listData, (size_t) count, sizeof ( struct formatRec ), (int (*) (const void*, const void*)) yearLformatSort ) ;
  else if ( mrropt == SMRR ) 
    qsort ( (void*) listData, (size_t) count, sizeof ( struct formatRec ), (int (*) (const void*, const void*)) smrrLformatSort ) ;
  else if ( mrropt == VMRR ) 
    qsort ( (void*) listData, (size_t) count, sizeof ( struct formatRec ), (int (*) (const void*, const void*)) vmrrLformatSort ) ;
}


int voteTitleKeyLseenCompare ( struct voteDbRec *search, struct voteDbRec *v1 )
{
  return ( search -> titleKey - v1 -> titleKey ) ;
}


int readLseenData ( struct formatRec listData [], struct voteDbRec *votes, TitleID nvotes, FILE *dbFp, int listId ) 
{
  int count = 0, len, i ;
  int noWithAttr, noWithoutAttr ;
  struct voteDbRec searchRec ;
  NameID nameKey ;

  nameKey = getName ( dbFp ) ;
  getFilmographyCounts ( dbFp, &noWithAttr, &noWithoutAttr ) ;
  for ( i = 0 ; i < noWithAttr ; i++ )
  {
    listData [ count ] . nameKey = nameKey ;
    listData [ count ] . titleKey = getTitle ( dbFp ) ;
    listData [ count ] . attrKey = getAttr ( dbFp ) ;
    if ( listId < NO_OF_CAST_LISTS )
    {
      len = getByte ( dbFp ) ;
      if ( len > 0 )
      (void) fseek ( dbFp, len, SEEK_CUR ) ;
    }
    listData [ count ] . name = NULL ;
    listData [ count ] . title = NULL ;
    listData [ count ] . attr = NULL ;
    listData [ count ] . year = 0 ;
    searchRec . titleKey = listData [ count ] . titleKey ;
    if ( bsearch ( (void*) &searchRec, (void*) votes, (size_t) nvotes, sizeof ( struct voteDbRec ), (int (*) (const void*, const void*)) voteTitleKeyLseenCompare )  != NULL )
      count++ ;
  }
  for ( i = 0 ; i < noWithoutAttr ; i++ )
  {
    listData [ count ] . nameKey = nameKey ;
    listData [ count ] . titleKey = getTitle ( dbFp ) ;
    if ( listId < NO_OF_CAST_LISTS )
    {
      len = getByte ( dbFp ) ;
      if ( len > 0 )
      (void) fseek ( dbFp, len, SEEK_CUR ) ;
    }
    listData [ count ] . attrKey = NOATTR ;
    listData [ count ] . name = NULL ;
    listData [ count ] . title = NULL ;
    listData [ count ] . attr = NULL ;
    listData [ count ] . year = 0 ;
    searchRec . titleKey = listData [ count ] . titleKey ;
    if ( bsearch ( (void*) &searchRec, (void*) votes, (size_t) nvotes, sizeof ( struct voteDbRec ), (int (*) (const void*, const void*)) voteTitleKeyLseenCompare )  != NULL )
     count++ ;
  }
  return ( count ) ;
}


int lseenCountSort ( struct lseenCount *r1, struct lseenCount *r2 )
{
  return ( r2 -> count - r1 -> count ) ;
}


void displayLseenCounts ( struct lseenCount counts[], int ncount, int mrropt, int yropt, struct voteDbRec *votes, TitleID nvotes, struct titleDbRec *years, TitleID nyears, struct mrrDbRec *ratings, TitleID nratings, FILE *dbFp, int listId ) 
{
  FILE *titleKeyFp = NULL, *titleIndexFp = NULL, *attrKeyFp = NULL, *attrIndexFp = NULL ;
  int i, count ;
  struct formatRec listData [ MAXFILMOGRAPHIES ] ;

  if ( mrropt != NONE || yropt != NONE )
  {
    titleKeyFp = openFile ( TITLEKEY ) ;
    titleIndexFp = openFile ( TITLEIDX ) ;
    attrKeyFp = openFile ( ATTRKEY ) ;
    attrIndexFp = openFile ( ATTRIDX ) ;
  }
  qsort ( (void*) counts, (size_t) ncount, sizeof ( struct lseenCount ), (int (*) (const void*, const void*)) lseenCountSort ) ;
  for ( i = 0 ; i < ncount ; i++ )
  {
    if ( mrropt != NONE || yropt != NONE ) 
    {
      (void) fseek ( dbFp, counts [ i ] . offset, SEEK_SET ) ;
      count = readLseenData ( listData, votes, nvotes, dbFp, listId ) ;
      mapLseenData ( listData, count, titleKeyFp, titleIndexFp, attrKeyFp, attrIndexFp ) ;
      sortLseenData ( listData, count, mrropt, yropt, years, nyears, ratings, nratings ) ;
      listData [ 0 ] . name = counts [ i ] . name ;
      displayLseenData ( listData, count, mrropt, yropt ) ;
      freeFormatData ( listData, count ) ;
    }
    else
      (void) printf ( "%3d %s\n", counts [ i ] . count, counts [ i ] . name ) ;
  }
  if ( mrropt != NONE || yropt != NONE )
  {
    (void) fclose ( titleKeyFp ) ;
    (void) fclose ( titleIndexFp ) ;
    (void) fclose ( attrKeyFp ) ;
    (void) fclose ( attrIndexFp ) ;
  }
}


void lseenList ( int listId, int mrropt, int yropt, int usopt, int threshold, struct voteDbRec *votes, TitleID nvotes, struct titleDbRec *years, TitleID nyears, struct  mrrDbRec *ratings, TitleID nratings )
{
  FILE *dbFp, *nameKeyFp, *nameIndexFp, *titleKeyFp = NULL, *titleIndexFp = NULL, *attrKeyFp = NULL, *attrIndexFp = NULL ;
  struct formatRec listData [ MAXFILMOGRAPHIES ] ;
  struct lseenCount *counts = NULL ;
  struct voteDbRec searchRec ;
  int i, len, ncount = 0, count ;
  char fn [ MAXPATHLEN ] ;
  NameID nameKey ;
  long offset ;
  int noWithAttr, noWithoutAttr ;

  if ( usopt )
  {
    titleKeyFp = openFile ( TITLEKEY ) ;
    titleIndexFp = openFile ( TITLEIDX ) ;
    attrKeyFp = openFile ( ATTRKEY ) ;
    attrIndexFp = openFile ( ATTRIDX ) ;
  }
  else
    if ( ( counts = calloc ( MAXLSEEN, sizeof ( struct lseenCount ) ) ) == NULL )
      moviedbError ( "out of memory" ) ;

  (void) constructFilename ( fn, filmographyDefs [ listId ] . stem, DBSEXT ) ;
  dbFp = openFile ( fn ) ;
  nameKeyFp = openFile ( NAMEKEY ) ;
  nameIndexFp = openFile ( NAMEIDX ) ;

  while ( TRUE )
  {
    count = 0 ;
    offset = ftell ( dbFp ) ;
    nameKey = getName ( dbFp ) ;
    if ( feof ( dbFp ) )
      break ;
    getFilmographyCounts ( dbFp, &noWithAttr, &noWithoutAttr ) ;
    for ( i = 0 ; i < noWithAttr ; i++ )
    {
      listData [ count ] . nameKey = nameKey ;
      listData [ count ] . titleKey = getTitle ( dbFp ) ;
      listData [ count ] . attrKey = getAttr ( dbFp ) ;
      if ( listId < NO_OF_CAST_LISTS )
      {
        len = getByte ( dbFp ) ;
        if ( len > 0 )
        (void) fseek ( dbFp, len, SEEK_CUR ) ;
      }
      listData [ count ] . name = NULL ;
      listData [ count ] . title = NULL ;
      listData [ count ] . attr = NULL ;
      listData [ count ] . year = 0 ;
      searchRec . titleKey = listData [ count ] . titleKey ;
      if ( bsearch ( (void*) &searchRec, (void*) votes, (size_t) nvotes, sizeof ( struct voteDbRec ), (int (*) (const void*, const void*)) voteTitleKeyLseenCompare )  != NULL )
        count++ ;
    }
    for ( i = 0 ; i < noWithoutAttr ; i++ )
    {
      listData [ count ] . nameKey = nameKey ;
      listData [ count ] . titleKey = getTitle ( dbFp ) ;
      if ( listId < NO_OF_CAST_LISTS )
      {
        len = getByte ( dbFp ) ;
        if ( len > 0 )
        (void) fseek ( dbFp, len, SEEK_CUR ) ;
      }
      listData [ count ] . attrKey = NOATTR ;
      listData [ count ] . name = NULL ;
      listData [ count ] . title = NULL ;
      listData [ count ] . attr = NULL ;
      listData [ count ] . year = 0 ;
      searchRec . titleKey = listData [ count ] . titleKey ;
      if ( bsearch ( (void*) &searchRec, (void*) votes, (size_t) nvotes, sizeof ( struct voteDbRec ), (int (*) (const void*, const void*)) voteTitleKeyLseenCompare )  != NULL )
      count++ ;
    }

    if ( count > 0 && count >= threshold )
      if ( !usopt ) 
      {
        counts [ ncount ] . count = count ;
        counts [ ncount ] . offset = offset ;
        counts [ ncount ] . nameKey = nameKey ;
        counts [ ncount ] . name = mapNameKeyToText ( nameKey, nameKeyFp, nameIndexFp ) ;
        if ( ++ncount > MAXLSEEN )
          moviedbError ( "out of space to store totals, use -min option" ) ;
      }
      else
      {
        mapLseenData ( listData, count, titleKeyFp, titleIndexFp, attrKeyFp, attrIndexFp ) ;
        sortLseenData ( listData, count, mrropt, yropt, years, nyears, ratings, nratings ) ;
        listData [ 0 ] . name = mapNameKeyToText ( nameKey, nameKeyFp, nameIndexFp ) ;
        displayLseenData ( listData, count, mrropt, yropt ) ;
        freeFormatData ( listData, count ) ;
      }
  }
  if ( ! usopt ) 
    displayLseenCounts ( counts, ncount, mrropt, yropt, votes, nvotes, years, nyears, ratings, nratings, dbFp, listId );

  (void) fclose ( dbFp ) ;
  (void) fclose ( nameKeyFp ) ;
  (void) fclose ( nameIndexFp ) ;
  if ( usopt )
  {
    (void) fclose ( titleKeyFp ) ;
    (void) fclose ( titleIndexFp ) ;
    (void) fclose ( attrKeyFp ) ;
    (void) fclose ( attrIndexFp ) ;
  }
}


int main ( int argc, char **argv )
{
  struct mrrDbRec   *ratings = NULL ;
  struct titleDbRec *years = NULL ;
  struct voteDbRec  *votes ;
  int    err = FALSE ;
  int    i, j, okopt ;
  int    mrropt = NONE ;
  int    yropt = NONE ;
  int    usopt = FALSE ;
  int    threshold = 0 ;
  int    listId = -1 ;
  TitleID    nyears = 0, nvotes = 0, nratings = 0 ;

  if ( argc == 1 )
    err = TRUE ;

  for ( i = 1 ; i < argc ; i++ )
    if ( strcmp ( "-mrr", argv [ i ] ) == 0 )
      mrropt = MRR ;
    else if ( strcmp ( "-smrr", argv [ i ] ) == 0 )
      mrropt = SMRR ;
    else if ( strcmp ( "-vmrr", argv [ i ] ) == 0 )
      mrropt = VMRR ;
    else if ( strcmp ( "-yr", argv [ i ] ) == 0 )
      yropt = YR ;
    else if ( strcmp ( "-yru", argv [ i ] ) == 0 )
      yropt = YRU ;
    else if ( strcmp ( "-us", argv [ i ] ) == 0 )
      usopt = TRUE ;
    else if ( strcmp ( "-min", argv [ i ] ) == 0 )
    {
      if ( ++i < argc )
        threshold = atoi ( argv [ i ] ) ;
      else
        err = TRUE ;
    }
    else
    {
      okopt = FALSE ;
      for ( j = 0 ; j < NO_OF_FILMOGRAPHY_LISTS ; j++ )
	if ( strcmp ( filmographyDefs [ j ] . option, argv [ i ] ) == 0 )
        {
	  okopt = TRUE ;
	  if ( listId != -1 )
	    moviedbError ( "lseen: only one database option allowed" ) ;
	  listId = j ;
	}
      if ( ! okopt )
      {
         (void) fprintf ( stderr, "lseen: unrecognised option %s\n", argv[i] ) ;
         err = TRUE ;
      }
    }

  if ( !err && listId == -1 )
  {
    (void) fprintf ( stderr, "lseen: no database specified\n" ) ;
    err = TRUE ;
  }

  if ( err )
    moviedbUsage ( LSEEN_USAGE1, LSEEN_USAGE2, NULL, NULL, NULL, NULL ) ;

  if ( ( votes = calloc ( MAXTITLES, sizeof ( struct voteDbRec ) ) ) == NULL )
    moviedbError ( "out of memory" ) ;
  nvotes = readVotesDb ( votes ) ;

  if ( mrropt != NONE )
  {
    nratings = ratingsDbSize ( ) ;
    if ( ( ratings = calloc ( nratings, sizeof ( struct mrrDbRec ) ) ) == NULL )
      moviedbError ( "out of memory" ) ;
    nratings = readRatingsDb ( ratings ) ;
  }

  if ( yropt != NONE )
  {
    if ( ( years = calloc ( MAXTITLES, sizeof ( struct titleDbRec ) ) ) == NULL )
      moviedbError ( "out of memory" ) ;
    nyears = readTitleDb ( years ) ;
  }

  lseenList ( listId, mrropt, yropt, usopt, threshold, votes, nvotes, years, nyears, ratings, nratings ) ;

  return ( 0 ) ;
}
